contracts/contracts/sbtc-withdrawal.clar

```diff
@@ -33,6 +45,110 @@
+    )
+      ;; Check that the caller is the current signer principal
+      (asserts! (is-eq (get current-signer-principal current-signer-data) tx-sender) ERR_INVALID_CALLER)
+
+      ;; Check whether it was already accepted or rejected
+      (asserts! (is-none (get status request)) ERR_ALREADY_PROCESSED)
+
+      ;; Check that fee is not higher than requesters max fee
+      (asserts! (<= fee requested-max-fee) ERR_FEE_TOO_HIGH)
+
+      ;; Burn the locked-sbtc
+      (try! (contract-call? .sbtc-token protocol-burn-locked (get amount request) requester))
+
+      ;; Mint the difference b/w max-fee of the request & fee actually paid back to the user in sBTC
+      (if (is-eq (- requested-max-fee fee) u0)
+        true
+        (try! (contract-call? .sbtc-token protocol-mint (- requested-max-fee fee) requester))
+      )
+
+      ;; Call into registry to confirm accepted withdrawal
+      (try! (contract-call? .sbtc-registry complete-withdrawal request-id true (some bitcoin-txid) (some signer-bitmap) (some output-index) (some fee)))
+
+      (ok true)
+  )
+)
+
+;; Reject a withdrawal request
+(define-public (reject-withdrawal (request-id uint) (signer-bitmap uint))
+  (let
+     (
+      (current-signer-data (contract-call? .sbtc-registry get-current-signer-data))   
+      (withdrawal (unwrap! (contract-call? .sbtc-registry get-withdrawal-request request-id) ERR_INVALID_REQUEST))
+     )
+
+    ;; Check that the caller is the current signer principal
+    (asserts! (is-eq (get current-signer-principal current-signer-data) tx-sender) ERR_INVALID_CALLER)
+
+    ;; Check that request status is currently-pending
+    (asserts! (is-none (get status withdrawal)) ERR_ALREADY_PROCESSED)
+
+    ;; Burn sbtc-locked & re-mint sbtc to original requester
+    (try! (contract-call? .sbtc-token protocol-unlock (get amount withdrawal) (get sender withdrawal)))
+
+    ;; Call into registry to confirm accepted withdrawal
+    (try! (contract-call? .sbtc-registry complete-withdrawal request-id false none none none none))
+
+    (ok true)
+  )
+)
+;; Reject multiple withdrawal requests
+(define-public (complete-withdrawals (withdrawals (list 100 {request-id: uint, status: bool, signer-bitmap: uint, bitcoin-txid: (optional (buff 32)), output-index: (optional uint), fee: (optional uint)})))
+  (let 
+      (
+          (current-signer-data (contract-call? .sbtc-registry get-current-signer-data))
+      )
+
+      ;; Check that the caller is the current signer principal
+      (asserts! (is-eq (get current-signer-principal current-signer-data) tx-sender) ERR_INVALID_CALLER)
+
+      (fold complete-individual-withdrawal-helper withdrawals (ok u0))
+  )
+)
+
+(define-private (complete-individual-withdrawal-helper (withdrawal {request-id: uint, status: bool, signer-bitmap: uint, bitcoin-txid: (optional (buff 32)), output-index: (optional uint), fee: (optional uint)}) (helper-response (response uint uint)))
+  (match helper-response 
+    index
```

Comment:
Hmmm, confused here? If *one* withdrawal fails they'll all fail. When that happens we want to know where the error is thrown so if the error message is already thrown we don't want to update it but return it.

Reply:
> If _one_ withdrawal fails they'll all fail. When that happens we want to know where the error is thrown so if the error message is already thrown we don't want to update it but return it.

Yup, that makes sense. If we were going to go this route, I think we should should just return a tuple with the error number and the index as distinct fields. 

But, I was thinking that the `complete-withdrawals` function would always try each withdrawal input and either:
1. Return the result, so return `(list 100 (optional uint))` (or something like that) where `none` means success and `(some uint)` is the error number; or
2. Return a tuple for the failures, so return `(list 100 {error_number: uint, index: uint})`.
 
The downside is that this would ideally lead to `(list 100 none)` (or whatever type you go with) being committed on chain, while the upside is that the signers wouldn't need to reprocess failures (since they would be making the same contract call with kinda the same inputs). If we short-circuit then there could be withdrawals after the failed entry that can succeed and the signers need to make another contract call.

Reply:
_Re: error messages_
We can change the return of the (ok response) to include a tuple but not sure I follow how returning a tuple for the error case would work? All error messages across the protocol currently return a uint, including this a similar list mechanic found in sbtc-deposit. I don't think breaking this convention makes sense.

_Re: processing w/ failures_
This is a great point that I think merits more thought since it'd refactor this mechanic in other places. I'd like to leave this as-is & open a discussion for further consideration, does that work?

Reply:
> _Re: error messages_ We can change the return of the (ok response) to include a tuple but not sure I follow how returning a tuple for the error case would work? All error messages across the protocol currently return a uint, including this a similar list mechanic found in sbtc-deposit. I don't think breaking this convention makes sense.

Eh, I'm more than okay with breaking this convention here (actually, is it only a convention or is it required?). But I think we can side-step this issue and just figure out if we should short-circuit.

> _Re: processing w/ failures_ This is a great point that I think merits more thought since it'd refactor this mechanic in other places. I'd like to leave this as-is & open a discussion for further consideration, does that work?

Great idea, let's leave this alone for now until we discuss it with everyone else.

Reply:
Uint as error messages are a universal convention (& requirement for a lot of trait contracts), not strictly a requirement. 

Copy, opened here: https://github.com/stacks-network/sbtc/discussions/261